### PR TITLE
Specify connexion for db upgrade

### PR DESCRIPTION
**Pull request by AurelienLavorel**
Specify connexion for db upgrade

### Description (*)
Compatibility with Magento Commerce split databases setup.

### Fixed Issues (if relevant)
1. https://github.com/magento/magento2/pull/16570#issuecomment-429830198

### Manual testing scenarios (*)
As before

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @AurelienLavorel. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by AurelienLavorel** [comment]
Hi @ihor-sviziev 
Could you launch the EE test please?
Thank you

**Comment by ihor-sviziev** [comment]
Hi @engcom-backlog-nazar,
Could you please check if the issue is fixed by this PR?

From my perspective it shouldn't work because it using only one connection, while we two  - each for different db connections. Unfortunately I can't check it on EE right now.

**Comment by AurelienLavorel** [comment]
I tried on a fresh setup (with sales and quote on other db) and I have no errors during upgrade, hope you'll have the same result.

**Comment by ihor-sviziev** [comment]
Hi @AurelienLavorel,
Nazar checked this PR and unfortunately it doesn't work.
Here is screenshot:
image


I see that issue related to different connections for different DBs, each connection don't have access to another DBs, only to own.
I think proper solution will be getting all order addresses as it was before applying https://github.com/magento/magento2/pull/16570 + load some small batches of data and process them + do not use objects, just manipulate with arrays.

For sure it will be slower that implementation from https://github.com/magento/magento2/pull/16570 for single-db installations, but it will be working solution for multi-db installations

**Comment by AurelienLavorel** [comment]
I can add a check if Magento_ScalableCheckout, Magento_ScalableInventory, or Magento_ScalableOms is enabled, we use old code version (with Magento collection) instead of the SQL query. Can you accept it?

**Comment by ihor-sviziev** [comment]
But it will be failing with big amount of quotes / orders data

**Comment by rikwillems** [comment]
Working with batches and doing some cleanup after each iteration helps a lot. It still works slow though. Also I see a lot of order addresses not being updated. Probably be due to quote cleanup in EE.

**Comment by ihor-sviziev** [comment]
Hi @AurelienLavorel,

> I can add a check if Magento_ScalableCheckout, Magento_ScalableInventory, or Magento_ScalableOms is enabled, we use old code version (with Magento collection) instead of the SQL query. Can you accept it?

I think yes, it will fix this issue and at least people without split DBs will have performance improvement.

Please update your PR accordingly

**Comment by AurelienLavorel** [comment]
I did that, thank you for your time

**Comment by rikwillems** [comment]
This seems a poor solution as it still blocks large setups with split database to upgrade. I'm working on a solution too.

**Comment by ihor-sviziev** [comment]
Hi @AurelienLavorel,
@rikwillems prepared PR https://github.com/magento/magento2/pull/18945 for fixing the same issue, will review both of them, and will decide what's better

**Comment by ihor-sviziev** [comment]
Hi @AurelienLavorel,
I reviewed PR https://github.com/magento/magento2/pull/18945 (and was re-created as https://github.com/magento/magento2/pull/19098), and it looks better, just contains few issues.

So I'm closing this PR as we have better proposition of fixing the same issue.

Thank you for your time!

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 1]
```diff
@@ -178,27 +187,54 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
-                $addressTable . '.parent_id = sales_order.entity_id',
-                ['quote_address_id' => 'quote_address.address_id']
-            )
-            ->joinInner(
-                ['quote_address' => $setup->getTable('quote_address')],
-                'sales_order.quote_id = quote_address.quote_id 
+        if (
+            $this->moduleManager->isEnabled('Magento_ScalableCheckout') ||
+            $this->moduleManager->isEnabled('Magento_ScalableInventory') ||
+            $this->moduleManager->isEnabled('Magento_ScalableOms')
```
Comment (orlangur) [inline app/code/Magento/Sales/Setup/UpgradeData.php:193]: Well, actually we cannot refer to Magento Commerce modules from Magento Open Source ones.

**Pull request closed without merging**